### DISCUSSION TITLE
Error running Shadow under tornettools after latest Fedora 37 update

### DISCUSSION THREAD

**Discussion by timwalsh300** [shadow/shadow#2686] [Q&A]
I'll preface this by acknowledging that the Shadow documentation only claims support for Fedora 34, 35, and 36. However, since 34 and 35 are now end-of-life, and we're approaching the release of 38, I think this is relevant. My Fedora 37 system received a large number of package updates, including critical security updates, yesterday evening. This morning, my tornettools simulations (including the minimal tornet-0.01 example) are now terminating with the following output. ~~The basic example Shadow simulations _without_ tornettools still seem to run fine, which is puzzling to me because tornettools appears to use Shadow in a straightforward manner under the hood (removing the few default arguments to Shadow didn't help either)~~. **The basic example Shadow simulations that don't use tor still seem to run fine, so it's a tor + Shadow issue instead of a tornettools issue.** I've already updated to the latest v2.3.0 of Shadow. Before I do something drastic like wiping this machine and starting over with Fedora 36, since that might also be affected, do we have any ideas?

```
2023-01-21 09:06:51 1674320811.648843 [tornettools] [INFO] Logging system initialized! Logging events to stdout and to '/home/tim/Projects/tornettools/tornet-0.01/tornettools.simulate.2023-01-21.09.06.51.log'
2023-01-21 09:06:51 1674320811.649026 [tornettools] [INFO] Seeded standard and numpy PRNGs with seed=1648442271
2023-01-21 09:06:51 1674320811.649126 [tornettools] [INFO] The argument namespace is: Namespace(do_version=False, quiet=0, seed=1648442271, command='simulate', tornet_config_path='/home/tim/Projects/tornettools/tornet-0.01', shadow_exe='/home/tim/.local/bin/shadow', shadow_args='--parallelism=4 --seed=666 --template-directory=shadow.data.template', shadow_config='shadow.config.yaml', do_compress=False, use_realtime=False, func=<function simulate at 0x7f203c7b1c60>, formatter_class=<class '__main__.CustomHelpFormatter'>, prefix='/home/tim/Projects/tornettools/tornet-0.01')
2023-01-21 09:06:51 1674320811.652242 [tornettools] [INFO] The platform is: Linux-6.1.6-200.fc37.x86_64-x86_64-with-glibc2.36
2023-01-21 09:06:51 1674320811.652381 [tornettools] [INFO] System info: uname_result(system='Linux', node='fedoraonframework', release='6.1.6-200.fc37.x86_64', version='#1 SMP PREEMPT_DYNAMIC Sat Jan 14 16:55:06 UTC 2023', machine='x86_64')
2023-01-21 09:06:51 1674320811.652617 [tornettools] [INFO] Starting a simulation from tornet prefix /home/tim/Projects/tornettools/tornet-0.01
2023-01-21 09:06:51 1674320811.652693 [tornettools] [INFO] Starting dstat
2023-01-21 09:06:51 1674320811.653114 [tornettools] [INFO] Starting free loop
2023-01-21 09:06:51 1674320811.657447 [tornettools] [INFO] Starting shadow
** Starting Shadow v2.3.0-274-gfba10618 2023-01-21--10:44:08 with GLib v2.74.1
thread 'shadow-worker' panicked at 'called `Result::unwrap()` on an `Err` value: ENOMEM', main/host/memory_manager/memory_mapper.rs:760:26
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'shadow-worker' panicked at 'called `Result::unwrap()` on an `Err` value: ENOMEM', main/host/memory_manager/memory_mapper.rs:760:26
thread 'shadow-worker' panicked at 'assertion failed: self.shim_shmem_lock.borrow().is_none()', main/host/host.rs:821:9
fatal runtime error: failed to initiate panic, error 5
thread 'shadow-worker' panicked at 'called `Result::unwrap()` on an `Err` value: ENOMEM', main/host/memory_manager/memory_mapper.rsthread ':shadow-worker760' panicked at ':assertion failed: self.shim_shmem_lock.borrow().is_none()26', main/host/host.rs:821:9
fatal runtime error: failed to initiate panic, error 2023-01-21 09:07:06 1674320826.691865 [tornettools] [INFO] Cleaning up
2023-01-21 09:07:06 1674320826.695631 [tornettools] [INFO] Done simulating; shadow returned code '-6'
2023-01-21 09:07:06 1674320826.695754 [tornettools] [ERROR] Shadow simulation did not complete successfully
```

**Comment by timwalsh300** [comment]
Update: I've found that the "Simple Tor Network" example in the Shadow documentation (https://shadow.github.io/docs/guide/getting_started_tor.html), not using tornettools, is producing the same error. I've tried rebuilding tor from the latest source, and building it from a much older source, and got the same error. All of these tor versions launch fine outside of Shadow. Here's the output from the "Simple Tor Network" example.

> ** Starting Shadow v2.3.0-274-gfba10618 2023-01-21--10:44:08 with GLib v2.74.1
thread 'shadow-worker' panicked at 'called `Result::unwrap()` on an `Err` value: ENOMEM', main/host/memory_manager/memory_mapper.rs:760:26
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'shadow-worker' panicked at 'assertion failed: self.shim_shmem_lock.borrow().is_none()', main/host/host.rs:821:9
fatal runtime error: failed to initiate panic, error 5
./run.sh: line 4: 106418 Aborted                 (core dumped) shadow --template-directory shadow.data.template tor-minimal.yaml > shadow.log

**Comment by stevenengler** [comment]
```
thread 'shadow-worker' panicked at 'called `Result::unwrap()` on an `Err` value: ENOMEM', main/host/memory_manager/memory_mapper.rs:760:26
```

Are you sure you have enough system memory? How much memory does your system have, and how much is remaining when you start the simulation? Are you running in Docker or any other virtualized/containerized environment? What happens if you run Shadow with the `--use-memory-manager false` flag? Have you followed any of the steps at https://shadow.github.io/docs/guide/system_configuration.html, specifically the ones about increasing the memory map counts?

**Comment by timwalsh300** [comment]
This machine has 32 GB of memory. Once I startup a 2% scale tornettools simulation, top shows me about 26 GB free and it slowly ticks down from there as the simulation progresses. This is a bare-metal installation of Fedora. **I just tried using the `--use-memory-manager false` flag, and the "Simple Tor Network" test completed successfully. I have one of my larger 2% scale tornettools simulations running OK now with the same flag applied.** I walked through everything on the System Configuration page, and I did find that map_max_count was set to 65530, so I made that change to 1073741824, but that didn't resolve the issue.

What are the consequences of continuing to use `--use-memory-manager false`?

**Reply by stevenengler** [reply]
Can you post the output of `df -h -t tmpfs`? And if it shows the space in `/dev/shm` being low, can you remount it to have more space? See #2581 and https://github.com/shadow/shadow/discussions/2622#discussioncomment-4484959 for details. Using `--use-memory-manager false` might make the simulation a little slower, but shouldn't have any effect on the simulation results.

**Reply by timwalsh300** [reply]
Here's that output. It looks like my simulation took almost twice as much time as usual, but worked as advertised. You're awesome, by the way.

> Filesystem      Size  Used Avail Use% Mounted on
tmpfs            16G     0   16G   0% /dev/shm
tmpfs           6.3G  2.0M  6.3G   1% /run
tmpfs            16G   28K   16G   1% /tmp
tmpfs           3.2G  144K  3.2G   1% /run/user/1000

**Reply by stevenengler** [reply]
@sporksmith Any idea why it would be returning `ENOMEM` with the memory manager enabled? 32 GB of memory and a 16 GB /dev/shm should be enough for the tor minimal test.

**Reply by sporksmith** [reply]
> Any idea why it would be returning ENOMEM with the memory manager enabled? 32 GB of memory and a 16 GB /dev/shm should be enough for the tor minimal test.

Yeah, that's weird. Seeing if I can reproduce in Docker...

> It looks like my simulation took almost twice as much time as usual,

Wow, that's surprising. Lately I've only seen a small % performance difference with it on or off, including local runs of the tor minimal test. Guess I should take a look at a bigger sim in #2581 . (Of course, it's also possible the difference is related to something else in fedora 38 or system configuration etc)

**Comment by sporksmith** [comment]
The minimal tor test works for me in a fedora 37 docker container. (I accidentally tried fedora 38, on which it turns out tgen doesn't build, but let's disregard that yak for now https://github.com/shadow/tgen/issues/44)

I had to make a small fix to the integration test scripts, which I'll send out shortly, but it's orthogonal to this.

Anyway this doesn't necessarily tell us a whole lot other than that we can't easily reproduce the issue. e.g. the problem could be related to the newer kernel, which wouldn't reproduce in Docker since it uses the host kernel.

**Reply by timwalsh300** [reply]
@sporksmith Yeah, it's the newer kernel. You gave me the idea to try booting with the previous 6.0.16 kernel, and Shadow's included "Simple Tor Network" test completed successfully (without disabling the memory manager). Then I booted back into the latest 6.1.6 kernel and it gave me the ENOMEM error again.

Now that I can run with or without the memory manager, I'll try to give you a more precise observation of the slowdown that I'm seeing with the memory manager disabled. I might have been comparing two simulations with different consensus files and a few other differences.

**Reply by sporksmith** [reply]
Filed https://github.com/shadow/shadow/issues/2692

**Reply by sporksmith** [reply]
I found the problem and a workaround (in case the performance difference of using `--memory-manager=false` is a significant issue for you). I expect we'll merge a proper fix within the next week or so. https://github.com/shadow/shadow/issues/2692#issuecomment-1402349631

**Reply by timwalsh300** [reply]
Very cool! I'm happy to keep running the 6.0.16 kernel for now to avoid the issue, but it's comforting to know that the issue might be fixed by the time Fedora removes this kernel.
